src/Spice86.Core/Emulator/OperatingSystem/DosFilePathResolver.cs

```diff
@@ -0,0 +1,172 @@
+namespace Spice86.Core.Emulator.OperatingSystem;
+
+using Spice86.Shared.Emulator.Errors;
+using Spice86.Shared.Interfaces;
+using Spice86.Shared.Utils;
+
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+/// <inheritdoc />
+public class DosFilePathResolver : IDosFilePathResolver {
+    private readonly ILoggerService _loggerService;
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    /// <param name="loggerService">The logger service implementation.</param>
+    public DosFilePathResolver(ILoggerService loggerService) => _loggerService = loggerService;
+    
+    /// <inheritdoc />
+    public string GetHostParentDirectory(string path) => Directory.GetParent(path)?.FullName ?? path;
+
+    /// <inheritdoc />
+    public string? GetActualCaseForFileName(string caseInsensitivePath) {
+        string? directory = Path.GetDirectoryName(caseInsensitivePath);
+        string? directoryCaseSensitive = GetDirectoryCaseSensitive(directory);
+        if (string.IsNullOrWhiteSpace(directoryCaseSensitive) || Directory.Exists(directoryCaseSensitive) == false) {
+            return null;
+        }
+        string realFileName = "";
+        string[] array = Directory.GetFiles(directoryCaseSensitive);
+        foreach (string file in array) {
+            string fileToUpper = file.ToUpperInvariant();
+            string searchedFile = caseInsensitivePath.ToUpperInvariant();
+            if (fileToUpper == searchedFile) {
+                realFileName = file;
+            }
+        }
+        if (string.IsNullOrWhiteSpace(realFileName) || File.Exists(realFileName) == false) {
+            return null;
+        }
+        return realFileName;
+    }
+
+    private static string? GetDirectoryCaseSensitive(string? directory) {
+        if (string.IsNullOrWhiteSpace(directory)) {
+            return null;
+        }
+        DirectoryInfo directoryInfo = new(directory);
+        if (directoryInfo.Exists) {
+            return directory;
+        }
+
+        if (directoryInfo.Parent == null) {
+            return null;
+        }
+
+        string? parent = GetDirectoryCaseSensitive(directoryInfo.Parent.FullName);
```

Comment:
It's a recursive call to itself, which can return null if:

```

if (string.IsNullOrWhiteSpace(directory)) {
            return null;
        }
[...]

        if (directoryInfo.Parent == null) {
            return null;
        }

        string? parent = GetDirectoryCaseSensitive(directoryInfo.Parent.FullName);
        if (parent == null) {
            return null;
        }
```

Furthermore, the compiler always considers it can be null if you mark anything nullable.

Reply:
I  refactored the parameter so it is now a string, instead of a nullable string.